### PR TITLE
NIFI-4726: Avoid concurrency issues in JoltTransformJSON

### PR DESCRIPTION
**Pull request by mgaido91**
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? NA 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly? NA
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly? NA
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties? NA

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered? NA

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by joewitt** [comment]
We should avoid the use of ThreadLocal storage completely.  NiFi will not give the same thread back to the same processor with any guarantees at all.  So we'd be caching/caching/caching and perhaps even leaking information across threads like that.  I think we should work the root cause angle with jolt

**Comment by mgaido91** [comment]
Thanks for your comment @joewitt. I think we can't fix this in jolt since the root cause is our concurrent usage of their utilities and fixing it there seems an overkill to me. What do you think about using a pool instead of the thread local then?

**Comment by joewitt** [comment]
@mgaido91 If the root problem appears to be that the jolt object needed is not thread safe then using a pool of lazily initialized objects makes great sense.  You can create them when one is not available and return it to the pool and that will naturally max out at total num threads for that processor.  You can clear out the pool whenever the processor is unscheduled.

Thanks!

**Comment by mgaido91** [comment]
Thanks for your suggestion @joewitt, I'll update this PR accordingly ASAP, thanks.

**Comment by ijokarumawak** [comment]
Hi @mgaido91 @joewitt , I was investigating the root cause of the issue. Finally, I think I was able to reproduce the ArrayIndexOutOfBoundsException with JoltTransformJson processor. You can find details in this Gist with a reproducible NiFi flow template. NOTE, this is not reproducible anymore with current NiFi 1.5.0 snapshot, you need to test it with older version of NiFi.
https://gist.github.com/ijokarumawak/65f51ea49029c1eafcc62c39e1a606bc

Does it seem reasonable? If that is the root cause we're looking for, then it doesn't caused by concurrent use of JsonUtils.

**Comment by joewitt** [comment]
Based on Koji's findings for the JIRA this appears to be resolved already by upgrading dependencies.  I recommend we close this @mgaido91

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/util/JsonUtilsPool.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import com.bazaarvoice.jolt.JsonUtil;
+import com.bazaarvoice.jolt.JsonUtilImpl;
+import org.apache.commons.pool.PoolableObjectFactory;
+import org.apache.commons.pool.impl.GenericObjectPool;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JsonUtilsPool {
+    private final static Logger LOG = LoggerFactory.getLogger(JsonUtilsPool.class);
+    static private GenericObjectPool pool;
```
Comment (joewitt) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/util/JsonUtilsPool.java]: dont believe we would want a static pool instance.  We can just make the pool a member of the processor.

**Comment by mgaido91** [comment]
sure, thanks @ijokarumawak and @joewitt. Sorry for the mistake and wasting your time.

**Comment by joewitt** [comment]
not a waste of time at all.  thanks for your efforts @mgaido91

**Pull request closed without merging**